### PR TITLE
Masq 191 hash keys

### PR DESCRIPTION
**Pull request by leventdem**
We want to hash the keys with a nonce: 
Each call to put/get with key, nonce => hash(nonce || key)

### CODE REVIEW

File: src/errors.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 const ERRORS = {
+  INVALID_ENCODING_FORMAT: 'Supporting encoding format: hex and base64',
```
Comment (sujeebant) [inline src/errors.js]: Supported encoding formats: hex and base64
OR Supporting encoding formats hex and base64

File: src/utils.js [hunk 2]
```diff
@@ -44,20 +44,38 @@ function dbExists (dbName) {
...
 /**
    * Get a value
    * @param {Object} db - The hyperDB instance
    * @param {CryptoKey} enckey - The enc/dec AES key
    * @param {string} key - The key name
+   * @param {string} nonce - The nonce as hex string
    * @returns {Promise}
    */
-const get = async (db, encKey, key) => {
+const get = async (db, encKey, key, nonce) => {
```
Comment (sujeebant) [inline src/utils.js]: You could reorder the parameters to have nonce after encKey

File: src/utils.js [hunk 3]
```diff
@@ -44,20 +44,38 @@ function dbExists (dbName) {
...
   if (!db) throw new MasqError(ERRORS.NO_DB)
   if (!encKey) throw new MasqError(ERRORS.NO_ENCRYPTION_KEY)
```
Comment (sujeebant) [inline src/utils.js:69]: Check if nonce argument is given

File: src/utils.js [hunk 4]
```diff
@@ -86,21 +112,24 @@ const list = async (db, encKey, prefix) => {
...
-  const decList = await Promise.all(list.map(async (elt) => ({
-    key: elt.key,
-    value: await decrypt(encKey, elt.value)
-  })))
+  const decList = await Promise.all(list.map(async (elt) => {
+    const dec = await decrypt(encKey, elt.value)
+    return {
+      key: dec.key,
+      value: dec.value
+    }
```
Comment (sujeebant) [inline src/utils.js]: Can't you return dec directly ?

File: src/utils.js [hunk 5]
```diff
@@ -86,21 +112,24 @@ const list = async (db, encKey, prefix) => {
...
+  }))
 
   const reformattedDic = decList.reduce((dic, e) => {
-    const el = Array.isArray(e) ? e[0] : e
```
Comment (sujeebant) [inline src/utils.js:100]: :+1:

File: src/utils.js (hunk 4 above)
Comment (G-Ray) [reply src/utils.js]: does dec contains more values than just `key` and `value` ?

File: src/crypto.js [hunk 6]
```diff
@@ -1,8 +1,15 @@
 import { ERRORS, checkObject, MasqError } from './errors'
 
-const genRandomBuffer = (len = 16) => {
+const _checkEncodingFormat = (format) => {
+  if (format !== 'hex' && format !== 'base64') throw new MasqError(ERRORS.INVALID_ENCODING_FORMAT)
+}
+
+const genRandomBuffer = (len = 16, encodingFormat) => {
```
Comment (G-Ray) [inline src/crypto.js]: The return output is inconsistent: it either return a Buffer, or a string.

I think we should keep this function simple by just generating a random buffer. The user of this function will call `toString(encoding)` if we wants

File: src/utils.js [hunk 7]
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
...
 }
 
+/**
+ *
+ * @param {string} key - The key (/prefix/key)
+ * @param {string} nonce - The nonce is used to ensure privacy and avoid de-anonymization (hex string)
+ */
+const hashKey = async (key, nonce) => {
+  if (key === '/' || key === '') return '/'
+  const prefixes = key.split('/').filter(prefix => prefix !== '')
```
Comment (G-Ray) [inline src/utils.js:54]: ok

File: src/utils.js [hunk 8]
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
...
+
+  let sanitizedKey = key
+
+  if (sanitizedKey[0] === '/') {
+    sanitizedKey = sanitizedKey.slice(1)
+  }
+
+  if (sanitizedKey[sanitizedKey.length - 1] === '/') {
+    sanitizedKey = sanitizedKey.slice(0, -1)
+  }
```
Comment (sujeebant) [inline src/utils.js:102]: Add a comment to explain that the key stored in the value is sanitized to remove the '/' at the beginning and end of the '/' separated key.

File: src/utils.js [hunk 9]
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
...
+/**
+ *
+ * @param {string} key - The key (/prefix/key)
+ * @param {string} nonce - The nonce is used to ensure privacy and avoid de-anonymization (hex string)
+ */
+const hashKey = async (key, nonce) => {
+  if (key === '/' || key === '') return '/'
+  const prefixes = key.split('/').filter(prefix => prefix !== '')
+  const hashedPrefixes = await Promise.all(prefixes.map(async (prefix) => hash256(`${nonce}${prefix}`)))
+  return hashedPrefixes.join('/')
```
Comment (G-Ray) [inline src/utils.js:56]: this will miss the leading / right ? is it an issue ?

File: src/utils.js [hunk 10]
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
...
     return {}
   }
-
-  const decList = await Promise.all(list.map(async (elt) => ({
-    key: elt.key,
-    value: await decrypt(encKey, elt.value)
-  })))
+  const decList = await Promise.all(list.map(async (elt) => {
+    const dec = await decrypt(encKey, elt.value)
+    return dec
```
Comment (sujeebant) [inline src/utils.js:134]: I think we can just use await Promise.all(list.map(elt => decrypt(encKey, elt.value)))

File: test/utils.js [hunk 11]
```diff
@@ -28,6 +29,14 @@ describe('MasqCommon utils', () => {
       const exists = await MasqCommon.utils.dbExists('test2')
       chai.assert.equal(exists, false, 'database must not exist')
     })
+
+    it('Should hash each subpaths of a given key', async () => {
+      const key = '/favoris/paris'
+      const prefixes = await MasqCommon.utils.hashKey(key, nonce)
+      const expected = 'ab595ffb49ed89da9e15a501d65f2fe0d0c47ef41ff346c2cd31246dd5c54003/e13e07a3ac6a83421551448b2ba94205a0fdaca27d622addc8bab44300f28290'
```
Comment (G-Ray) [inline test/utils.js:40]: I think we should keep the leading '\', what to you think ?

File: src/utils.js (hunk 8 above)
Comment (G-Ray) [reply src/utils.js:102]: ~~IIUC this sanitization step is not needed, as `hashKey()` will remove unnecessary slashes anyway ?~~

File: test/crypto.js [hunk 12]
```diff
@@ -2,215 +2,233 @@
...
-      const enc = await MasqCommon.crypto.encrypt(cryptokey, data)
+      const enc = await crypto.encrypt(cryptokey, data)
       chai.assert.exists(enc.iv, 'iv must exist')
       chai.assert.exists(enc.ciphertext, 'ciphertext must exist')
 
       // Just to be sure that everything is working well.
-      const sameMK = await MasqCommon.crypto.decryptMasterKey(passphrase, protectedMK)
-      const sameCryptokey = await MasqCommon.crypto.importKey(sameMK)
-      const dec = await MasqCommon.crypto.decrypt(sameCryptokey, enc)
+      const sameMKAndNonce = await crypto.decryptMasterKeyAndNonce(passphrase, protectedMK)
```
Comment (G-Ray) [inline test/crypto.js:77]: don't hesitate to do `const { masterKey } = await...`

File: test/crypto.js [hunk 13]
```diff
@@ -2,215 +2,233 @@
 /* global MasqCommon */
 /* global chai */
 
+const { ERRORS } = MasqCommon.errors
+const { crypto } = MasqCommon
```
Comment (G-Ray) [inline test/crypto.js:6]: +1

File: test/utils.js [hunk 14]
```diff
@@ -39,37 +48,73 @@ describe('MasqCommon utils', () => {
...
+    it('Should list (export) keys/values for a prefix with / or not at the beginning/end', async () => {
+      const masterKey = await MasqCommon.crypto.genAESKey()
+      const prefix1 = ('/fav/')
+      const prefix2 = ('/fav')
+      const prefix3 = ('fav')
+      const item = { one: '1' }
+      const db = await MasqCommon.utils.createPromisifiedHyperDB('dB9')
+      await MasqCommon.utils.dbReady(db)
+      await MasqCommon.utils.put(db, masterKey, nonce, `${prefix1}one`, item)
+      const list1 = await MasqCommon.utils.list(db, masterKey, nonce, prefix1)
```
Comment (G-Ray) [inline test/utils.js]: maybe we can verify that the retrieved item in the list match the exepected one ?

File: src/crypto.js [hunk 15]
```diff
@@ -1,10 +1,22 @@
...
 const genRandomBuffer = (len = 16) => {
   const values = window.crypto.getRandomValues(new Uint8Array(len))
   return Buffer.from(values)
 }
 
+const genRandomBufferAsStr = (len = 16, encodingFormat) => {
+  if (encodingFormat) {
+    _checkEncodingFormat(encodingFormat)
+  }
+  const values = window.crypto.getRandomValues(new Uint8Array(len))
```
Comment (G-Ray) [inline src/crypto.js]: you should reuse your function `genRandomBuffer` here :)

File: src/crypto.js [hunk 16]
```diff
@@ -1,10 +1,22 @@
...
+  return encodingFormat ? Buffer.from(values).toString(encodingFormat) : Buffer.from(values)
```
Comment (G-Ray) [inline src/crypto.js]: Once you get the buffer with `genRandomBuffer`, return a string, an use a default encoding if none is specified

File: src/crypto.js (hunk 15 above)
Comment (leventdem) [reply src/crypto.js]: yes, right.

File: src/crypto.js [hunk 17]
```diff
@@ -1,10 +1,22 @@
...
   const values = window.crypto.getRandomValues(new Uint8Array(len))
   return Buffer.from(values)
 }
 
+const genRandomBufferAsStr = (len = 16, encodingFormat = 'hex') => {
+  if (encodingFormat) {
+    _checkEncodingFormat(encodingFormat)
+  }
+  const buf = genRandomBuffer(len)
+  return buf.toString(encodingFormat)
```
Comment (G-Ray) [inline src/crypto.js:17]: :+1:

File: test/crypto.js (hunk 13 above)
Comment (G-Ray) [reply test/crypto.js:6]: if you have other problems with globals, you should be able to declare these var locally inside the describe section

**Review by G-Ray — APPROVED** [review]

**Comment by G-Ray** [comment]
We should set the key, dek and nonce only once, to avoid passing it around each function `get` `put` etc...

File: src/utils.js (hunk 7 above)
Comment (G-Ray) [reply src/utils.js]: what if key is `//mykey//` ?

File: test/utils.js [hunk 18]
```diff
@@ -19,109 +24,182 @@ describe('MasqCommon utils', () => {
...
     it('Should put/get encrypted values in the db', async () => {
-      const masterKey = await MasqCommon.crypto.genAESKey()
+      const masterKey = await crypto.genAESKey()
       const item = { one: '1' }
-      const db = await MasqCommon.utils.createPromisifiedHyperDB('dB2')
-      await MasqCommon.utils.dbReady(db)
-      await MasqCommon.utils.put(db, masterKey, 'one', item)
-      const dec = await MasqCommon.utils.get(db, masterKey, 'one')
+      const keyName = 'one'
+      // need to check if encryption is ok
```
Comment (G-Ray) [inline test/utils.js]: need to check ?

Comment (leventdem) [reply test/utils.js]: no old comment

File: src/utils.js (hunk 7 above)
Comment (sujeebant) [reply src/utils.js:54]: Why filter all empty strings ? is a key like '/abc//def' valid ?

**Review by sujeebant — APPROVED** [review]

**Pull request merged**